### ISSUE TITLE
[JENKINS-14108] Multiple parameterized upstreams with joined in one "blocked while.." job result in multiple executions rather one only

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/parameterized-trigger-plugin#530]
<p>Job setup:<br/>
TestRoot - TestP1 - TestJoin<br/>
&#45; TestP2 /</p>

<p>All of the jobs are configurated with parameterized string parameters in pre job and TestRoot, TestP1 and TestP2 uses "Trigger parameterized build on other projects" with file as input.<br/>
The TestJoin is configurated with "Block while upstreams are build"</p>

<p>The issue is as follows:<br/>
The Jenkins build-in feature works ok and does not execute while jobs are building upstream, but the big problem is that the parameterized trigger kicks off two job executions(or as many as there are parallel upstream) rather than just one.</p>

<p>It has been fine one this server when it was original configurated. At some point I moved some upstreams to join in a separated joiner. It seems that changing of configuration triggered the problem. The "funny" thing is that the problem occurs not only on the job I changed configuration but on all using the this setup. </p>

<p>The test setup mentioned above is created after the problem occured and it still have the problem.</p>

<p>I have updated the plugin to get the latest release, but it did not help on this problem.</p>

<p>This is showstopper for me, as I cannot have more than one job execution due to the fact that I need one joined result of my whole "graph" and then continue further actions based on this..<br/>
I have not seen it earlier in my use of the this plugin.</p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=bic_schneider">bic_schneider</a>, imported from: <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-14108" target="_blank">Multiple parameterized upstreams with joined in one "blocked while.." job result in multiple executions rather one only</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=huybrechts">huybrechts</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Blocker
<li><b>component(s)</b>: parameterized-trigger-plugin
<li><b>label(s)</b>: jenkins, plugin, plugins
<li><b>resolution</b>: Won't Fix
<li><b>resolved</b>: 2014-09-14T14:02:39+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 4
<li><b>imported</b>: 2025-12-07
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>Job setup:
TestRoot - TestP1 - TestJoin
        &#45; TestP2 /</p>

<p>All of the jobs are configurated with parameterized string parameters in pre job and TestRoot, TestP1 and TestP2 uses "Trigger parameterized build on other projects" with file as input.
The TestJoin is configurated with "Block while upstreams are build"</p>

<p>The issue is as follows:
The Jenkins build-in feature works ok and does not execute while jobs are building upstream, but the big problem is that the parameterized trigger kicks off two job executions(or as many as there are parallel upstream) rather than just one.</p>

<p>It has been fine one this server when it was original configurated. At some point I moved some upstreams to join in a separated joiner. It seems that changing of configuration triggered the problem. The "funny" thing is that the problem occurs not only on the job I changed configuration but on all using the this setup. </p>

<p>The test setup mentioned above is created after the problem occured and it still have the problem.</p>

<p>I have updated the plugin to get the latest release, but it did not help on this problem.</p>

<p>This is showstopper for me, as I cannot have more than one job execution due to the fact that I need one joined result of my whole "graph" and then continue further actions based on this..
I have not seen it earlier in my use of the this plugin.</p></pre>
</details>
</details>
<details><summary><i>environment</i></summary>

```
Linux: <br/>
Jenkins release: 1.449
```
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>bic_schneider:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-14108?focusedId=164032&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-164032">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>additional information:
The plugin/build in feature: "Build other projects" work fine concerning this issue.. Eg it only generate one build after "Block while upstreams..." has released the block.</p>
</pre>
</details>

</ul>
</details>

<p>additional information:<br/>
The plugin/build in feature: "Build other projects" work fine concerning this issue.. Eg it only generate one build after "Block while upstreams..." has released the block.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>cjo9900:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-14108?focusedId=166170&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-166170">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Can you add some details on the config so this can be reproduced.</p>
</pre>
</details>

</ul>
</details>

<p>Can you add some details on the config so this can be reproduced.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>cjo9900:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-14108?focusedId=166482&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-166482">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I have talked to Claus on the cause of this issue and found that the behaviour is caused by how the core handled the dependencies. </p>

<p>The use case was 
Job A param trigger -&gt; Job B1 Param trigger -\
                    \&gt; Job B2 Param trigger -&gt; Job C</p>

<p>and parameters from jobs B1 and B2 passed to Job C are different. </p>

<p>In older versions before Jenkins ~1.402, job C would be triggered only once so would imitate a join of B1 and B2. After 1.402 Job C would be triggered twice once with the parameters from Job B1 and once with parameters from Job B2 as these are different.</p>

<p>In this case the join plugin cannot be used as the downstream job C required parameters from B1 and B2, and links upstream started that the join plugin cannot pass.</p>

<p>Possible solution is to have a special ParameterAction that implements the QueueAction(returning false even if there are existing parameters) and FoldableAction interface allowing parameters to be combined into a single Action and build. Which should then allow this join behaviour to occur.</p>

<p>However this may cause issues for other plugins wanting to get the ParametersAction.</p>
</pre>
</details>

</ul>
</details>

<p>I have talked to Claus on the cause of this issue and found that the behaviour is caused by how the core handled the dependencies. </p>

<p>The use case was <br/>
Job A param trigger -> Job B1 Param trigger -\<br/>
    \> Job B2 Param trigger -> Job C</p>

<p>and parameters from jobs B1 and B2 passed to Job C are different. </p>

<p>In older versions before Jenkins ~1.402, job C would be triggered only once so would imitate a join of B1 and B2. After 1.402 Job C would be triggered twice once with the parameters from Job B1 and once with parameters from Job B2 as these are different.</p>

<p>In this case the join plugin cannot be used as the downstream job C required parameters from B1 and B2, and links upstream started that the join plugin cannot pass.</p>

<p>Possible solution is to have a special ParameterAction that implements the QueueAction(returning false even if there are existing parameters) and FoldableAction interface allowing parameters to be combined into a single Action and build. Which should then allow this join behaviour to occur.</p>

<p>However this may cause issues for other plugins wanting to get the ParametersAction.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>ikedam:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-14108?focusedId=207560&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-207560">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>This looks a current designated behavior of Jenkins.
When triggerd a job with different parameters, Jenkins should trigger builds for each parameters, and shouldn't merge them.</p>

<p>It's recommended to use <a href="https://wiki.jenkins-ci.org/display/JENKINS/Promoted+Builds+Plugin" class="external-link" target="_blank" rel="nofollow noopener">Promoted Builds plugin</a> or <a href="https://wiki.jenkins-ci.org/display/JENKINS/Join+Plugin" class="external-link" target="_blank" rel="nofollow noopener">Join plugin</a> for this scenario.</p>

<p>Anyway, this issue is very old, and I'm afraid there are no longer users interested in this issue. (And I don't think this is still a blocker.)
I plan to close this issue with "Won't Fix" in a month.
Please let me know if anyone still looks this issue.</p>
</pre>
</details>

</ul>
</details>

<p>This looks a current designated behavior of Jenkins.<br/>
When triggerd a job with different parameters, Jenkins should trigger builds for each parameters, and shouldn't merge them.</p>

<p>It's recommended to use <a href="https://wiki.jenkins-ci.org/display/JENKINS/Promoted+Builds+Plugin" class="external-link" target="_blank" rel="nofollow noopener">Promoted Builds plugin</a> or <a href="https://wiki.jenkins-ci.org/display/JENKINS/Join+Plugin" class="external-link" target="_blank" rel="nofollow noopener">Join plugin</a> for this scenario.</p>

<p>Anyway, this issue is very old, and I'm afraid there are no longer users interested in this issue. (And I don't think this is still a blocker.)<br/>
I plan to close this issue with "Won't Fix" in a month.<br/>
Please let me know if anyone still looks this issue.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>ikedam:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-14108?focusedId=210468&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-210468">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>It's the current designated behavior of Jenkins.</p>
</pre>
</details>

</ul>
</details>

<p>It's the current designated behavior of Jenkins.</p>
